File: pa1/List.java
/* JULIAN TO (jcto)
 * PA 1
 * CS 101
 * PROFESSOR TANTALO
 * SPRING 2019
 */

//node object that forms the linked list
class Node {
	int data;
	int position;
	Node next;
	Node prev;

	// constructor: takes the first element in the array and stores it in the node
	public Node(int input) {
		this.data = input;
		// this.position = 0;
		this.next = null;
		this.prev = null;
	}
}

//Doubly-Linked List for use with Lex.java
class List {
	// front of the list
	Node head;
	// end of the list
	Node tail;
	// cursor element
	Node cursor;

	// creates a new LinkedList
	public List() {
		// front = new Node(piece);
		head = null;
		tail = null;
		cursor = null;
	}

	// returns the length of the list, 0 if empty
	public int length() {
		int length = 0;
		// if head is null, list must be empty
		if (head == null) {
			return length;
		}
		Node start = head;
		length++; // something is in the list, so it must be at least one long
		while (start.next != null) {
			start = start.next;
			length++;
		}
		return length;
	}

	// returns the index (position) of the cursor element
	public int index() {
		int cursorTrack = -1;
		// default state
		Node cursorRadar = cursor;
		if (cursorRadar == null) {
			return -1;
		}

		Node start = head;
		while (start != null) {
			start = start.next;
			cursorTrack++;
		}
		cursorTrack++;
		return cursorTrack;
	}

	// like the above function, but returns the element instead
	public int get() {
		if (this.length() <= 0 || this.index() <= 0) {
			throw new RuntimeException("LIST EMPTY OR CURSOR UNDEFINED");
		}
		// return this.cursor.position;
		Node start = head;
		// int position = 0;
		while (start != cursor) {
			start = start.next;
		}
		return start.data;
	}

	// returns the int at the front of the list, else returns null
	public int front() {
		if (this.length() <= 0) {
			throw new IllegalArgumentException("Length of List is zero!");
		}
		return head.data;
	}

	// traverse the list to the end and then get the end returns the string
	// null if length is zero or less
	public int back() {
		if (this.length() <= 0) {
			throw new IllegalArgumentException("Length of List is zero!");
		}
		Node start = head;
		while (start.next != null) {
			start = start.next;
		}
		return start.data;
	}

	// checks if two lists are equal
	boolean equals(List L) {
		// used to traverse the array and compare values
		Node start = head;
		Node start2 = L.head;
		// checks for same length
		if (this.length() != L.length()) {
			return false;
		}
		//two empty lists are equal, I presume
		if(this.length() == 0 && L.length() == 0)
		{
			return true;
		}
		while (start.next != null || start2.next != null) {
			start = start.next;
			start2 = start2.next;
			// checks for equality of elements
			if (start.data != start2.data) {
				return false;
			}
		}
		return true;
	}

	// List Manipulation
	// clears the list
	void clear() {
		// just set the head to null cause lazy
		this.head = null;
		this.tail = null;
		this.cursor = null; //cursor is null too
	}

	// puts cursor under front element
	void moveFront() {
		if (this.length() <= 0) {
			// do nothing!
		} else {
			cursor = head;
		}

	}

	// puts cursor under rear element
	void moveBack() {
		if (this.length() <= 0) {
			// throw new IllegalArgumentException("Length is zero!");
			// DO NOTHING
		} else {
			Node start = head;
			while (start.next != null) {
				start = start.next;
			}
			cursor = start;
		}
	}

	// puts cursor on the element one previous to the current
	void movePrev() {
		// if cursor at the front make it undefined
		if (cursor == head) {
			cursor = null;
		}
		// if cursor is undefined
		else if (this.index() == -1) {
			// do nothing
		} else {
			cursor = cursor.prev;
		}
	}

	// puts cursor on the element one after the current
	void moveNext() {
		// if cursor at the end, make it undefined too
		if (cursor.next == null) {
			cursor = null;
		}
		// if cursor is undefined
		else if (this.index() == -1) {
			// do nothing
			// System.out.println("am i here?");
		} else {
			cursor = cursor.next;
		}
	}

	// adds a node to the front of the list
	void prepend(int data) {
		// makes a new node
		Node freshNode = new Node(data);
		// if list is empty, make it the first element
		if (this.length() == 0) {
			head = freshNode;
			tail = freshNode;
		}
		// if list is non empty, put the element at the front
		else {
			// inserts into the front by making the current head point to the inserted node
			freshNode.next = head;
			this.head.prev = freshNode;
			// then, set head to the new Node
			head = freshNode;
		}
	}

	// adds a node to the end of the list
	void append(int data) {
		// make a new node
		Node freshNode = new Node(data);
		if (this.length() == 0) {
			// System.out.println("me append");
			head = freshNode;
			tail = freshNode;
		}
		// if not empty, traverse to the end and add to it
		else {
			// System.out.println("me append2");
			// pretty much the opposite of what happened above
			tail.next = freshNode;
			freshNode.prev = tail;
			tail = freshNode;
		}
	}

	// adds an element before the current cursor's position
	void insertBefore(int data) {
		// check for prereqs
		if (this.length() <= 0 || this.index() < 0) {
			throw new IllegalArgumentException("Cursor undefined or invalid List!");
		}
		Node freshNode = new Node(data);
		// if the cursor is on the head, make the inserted Node the head
		if (cursor == head) {
			// System.out.println("test");
			head.prev = freshNode;
			freshNode.next = head;
			// check if the list is length 1
			if (head == tail) {
				tail = head;
			}
			head = freshNode;
		}
		// otherwise insert normally
		else {
			Node temp = cursor.prev;
			cursor.prev.next = freshNode;
			cursor.prev = freshNode;

			freshNode.next = cursor;
			freshNode.prev = temp;
		}
	}

	// adds an element after the current cursor's position
	void insertAfter(int data) {
		if (this.length() <= 0 || this.index() < 0) {
			throw new IllegalArgumentException("Cursor undefined or invalid List!");
		}
		Node freshNode = new Node(data);
		// if the cursor is on the tail, make the inserted Node the tail
		if (cursor == tail) {
			// System.out.println("test");
			tail.next = freshNode;
			freshNode.prev = tail;
			// check if the list is length 1
			if (head == tail) {
				// head = freshNode.prev;
			}
			tail = freshNode;
		} else {
			// otherwise insert normally
			Node temp = cursor.next;
			cursor.next.prev = freshNode;
			cursor.next = freshNode;

			freshNode.prev = cursor;
			freshNode.next = temp;
		}
	}

	// delete the front element
	void deleteFront() {
		if (this.length() == 0) {
			throw new IllegalArgumentException("Length is zero!");
		}
		// if length 1, just delete the whole list
		if (this.length() == 1) {
			head = null;
			tail = null;
			return;
		}
		// simply move the head to the next element
		head.next.prev = null;
		head = head.next;
	}

	// delete the rear element
	void deleteBack() {
		if (this.length() == 0) {
			throw new IllegalArgumentException("Length is zero!");
		}
		// if length 1, just delete the whole list
		if (this.length() == 1) {
			head = null;
			tail = null;
			return;
		}
		// delete the tail
		tail.prev.next = null;
		tail = tail.prev;
	}

	// delete the cursor element
	void delete() {
		Node point = cursor;
		// delete the links in the list that connect cursor to the rest of the list
		if (point == head) {
			head = cursor.next;
			head.prev = null;
			cursor = null;
			return;
		}
		if (point == tail) {
			// System.out.println("here?");
			tail = cursor.prev;
			tail.next = null;
			cursor = null;
			return;
		}
		if (point != head && point != tail) {
			cursor.prev.next = point.next;
			cursor.next.prev = point.prev;
			cursor = null;
			return;
		}
	}

	// toString method for printing
	public String toString() {
		Node start = head;
		String output = "";
		// if list is empty, print nothing
		if (this.length() == 0) {
			// System.out.println("fugg");
			return "";
		}
		while (start.next != null) {
			// System.out.println(start.data);
			output = output + start.data + " ";
			start = start.next;
		}
		output = output + start.data + " ";
		return output;
	}

	// copy the list into a new list
	public List copy() {
		List copiedList = new List();
		//if length of current list is zero, do nothing
		if(this.length() == 0)
		{
			return copiedList;
		}
		Node start = head;
		// traverse the list, and copy the elements into a new list
		while (start.next != null) {
			copiedList.append(start.data);
			start = start.next;
		}
		copiedList.append(start.data);
		return copiedList;
	}

	// joins two lists together, with the input List being on the end
	// note: not tested
	List concat(List L) {
		// List newList = new List();
		tail.next = L.head;
		L.head.prev = tail;

		// change the tail to L's tail
		tail = L.tail;
		return this;
	}
}

File: pa2/List.c
/*
JULIAN TO (jcto)
pa2: CMPS 101
Professor Tantalo
Spring 2019
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "List.h"
// STRUCTS

// defines the Node which holds all data
typedef struct NodeItem {
  struct NodeItem *nextItem;
  struct NodeItem *prevItem;
  int data;
} NodeItem;

// save some typing
typedef NodeItem *Node;

// defines the List which holds all the Nodes
typedef struct ListObject {
  Node head;   // defines the head of the list
  Node tail;   // defines the tail of the list
  Node cursor; // defines the cursor object

  // other data points
  // int index;
  // int length;

} ListObject;

// makes a new Nodes
Node newNode(int data) {
  NodeItem *x = malloc(sizeof(NodeItem));
  x->data = data;
  x->nextItem = NULL;
  x->prevItem = NULL;
}

// frees a Node from memory
void freeNode(Node *n) {
  if (*n != NULL && n != NULL) {
    free(*n);
    *n = NULL;
  }
}

// makes a new list, returning the LIst
List newList(void) {
  List L;
  L = malloc(sizeof(ListObject));
  // set head, tail, and cursor all to null
  L->head = NULL;
  L->tail = NULL;
  L->cursor = NULL;
  // index defaults to -1, length is
  // new_List->index = -1;
  // new_List->length = 0;
  return L;
}

// MOST FUNCTIONS TAKE IN THE LIST AS AN ARGUMENT FOR THE FUNCTION

// returns the length of the list, returns zero if empty
int length(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  Node pointer = L->head;
  int length = 0;
  // if head is null, list is probably empty (barring weird circumstance)
  if (pointer == NULL) {
    return length;
  }
  length++;
  while (pointer->nextItem != NULL) {
    pointer = pointer->nextItem;
    length++;
  }
  return length;
}

// returns the index (position) of the cursor element
int index(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  int cursorTracker = -1;
  // if cursor undefined, return -1
  Node cursorRadar = NULL;
  cursorRadar = L->head;
  if (cursorRadar == NULL) {
    return cursorTracker;
  }
  // otherwise, traverse the list until cursor is located
  NodeItem *pointer = L->head;
  while (pointer->nextItem != NULL) {
    if (pointer == cursorRadar) {
      break;
    }
    pointer = pointer->nextItem;
    cursorTracker++;
  }
  cursorTracker++;
  return cursorTracker;
}

// like the above function, but returns the value of the cursor instead
int get(List L) {
  // for undefined list
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // for lists with no cursor
  if (L->cursor == NULL) {
    printf("\nERROR: cursor not defined");
    exit(1);
  }
  NodeItem *pointer = L->head;
  while (pointer->nextItem != NULL) {
    if (pointer == L->cursor) {
      break;
    }
    pointer = pointer->nextItem;
  }
  return pointer->data;
}

// returns the data point at the front of the LIST
int front(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  int listLength = length(L);
  // if length is zero, throw error
  if (listLength == 0) {
    printf("\nERROR: LIST LENGTH IS ZERO\n");
    exit(1);
  }
  return L->head->data;
}

// returns the data point at the end of the LIST
int back(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  int listLength = length(L);
  // if length is zero, throw error
  if (listLength == 0) {
    printf("\nERROR: LIST LENGTH IS ZERO\n");
    exit(1);
  }
  return L->tail->data;
}

// check if two lists are equal/the same
int equals(List A, List B) {
  // idiot-proofing
  if (A == NULL || B == NULL) {
    return 0;
  }
  // traverse-rs
  Node start = A->head;
  Node start2 = B->head;
  // two lists of different lengths can't be equal
  if (length(A) != length(B)) {
    return 0;
  }
  // traverse and check each nodeItem
  while (start->nextItem != NULL && start2->nextItem != NULL) {
    if (start->data != start2->data) {
      return 0;
    }
    start = start->nextItem;
    start2 = start2->nextItem;
  }
  // check the last item on each list
  if (start->data != start2->data) {
    return 0;
  }
  return 1;
}

// manipulation FUNCTIONS
void clear(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  L->head = NULL;
  L->tail = NULL;
  L->cursor = NULL;
}

// put cursor under front elements
void moveFront(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  if (length(L) <= 0) {
    return; // do nothing
  } else {
    L->cursor = L->head;
  }
}

// put cursor under front elements
void moveBack(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  if (length(L) <= 0) {
    return; // do nothing
  } else {
    L->cursor = L->tail;
  }
}

// move cursor one back from the current point
void movePrev(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // if cursor at the front of the list, make cursor UNDEFINED
  if (L->cursor == L->head) {
    L->cursor = NULL;
  }
  // if cursor is undefined, do nothing
  if (L->cursor == NULL) {
    return;
  } else {
    L->cursor = L->cursor->prevItem;
  }
}

// move cursor one forward from the current point
void moveNext(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // if cursor at the front of the list, make cursor UNDEFINED
  if (L->cursor == L->head) {
    L->cursor = NULL;
  }
  // if cursor is undefined, do nothing
  if (L->cursor == NULL) {
    return;
  } else {
    L->cursor = L->cursor->nextItem;
  }
}

// adds a node to the front of the list
void prepend(List L, int data) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // make a new Node
  Node freshNode = newNode(data);
  // if list is empty, make this the first node
  if (length(L) == 0) {
    L->head = freshNode;
    L->tail = freshNode;
  }
  // otherwise make this the new head
  else {
    freshNode->nextItem = L->head;
    L->head->prevItem = freshNode;
    L->head = freshNode;
  }
}

// adds a node to the end of the list
void append(List L, int data) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // make a new Node
  Node freshNode = newNode(data);
  // if list is empty, make this the first node
  if (length(L) == 0) {
    L->head = freshNode;
    L->tail = freshNode;
  }
  // otherwise make this the new tail
  else {
    freshNode->prevItem = L->tail;
    L->tail->nextItem = freshNode;
    L->tail = freshNode;
  }
}

// adds an element before the current cursor's position
void insertBefore(List L, int data) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // check for valid arguments
  if (length(L) <= 0 || index(L) < 0) {
    printf("ERROR: Invalid List or Cursor undefined\n");
    exit(1);
  }
  Node freshNode = newNode(data);
  // if cursor is on the head, make the new Node the head
  if (L->cursor == L->head) {
    L->head->prevItem = freshNode;
    freshNode->nextItem = L->head;
    L->head = freshNode;
  }
  // otherwise insert normally
  else {
    Node temp = L->cursor->prevItem;
    L->cursor->prevItem->nextItem = freshNode;
    L->cursor->prevItem = freshNode;

    freshNode->nextItem = L->cursor;
    freshNode->prevItem = temp;
  }
}

void insertAfter(List L, int data) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  // check for valid arguments
  if (length(L) <= 0 || index(L) < 0) {
    printf("ERROR: Invalid List or Cursor undefined\n");
    exit(1);
  }
  Node freshNode = newNode(data);
  // if cursor is on the head, make the new Node the head
  if (L->cursor == L->tail) {
    L->tail->nextItem = freshNode;
    freshNode->prevItem = L->tail;
    L->tail = freshNode;
  }
  // insert normally otherwise
  else {
    Node temp = L->cursor->nextItem;
    L->cursor->nextItem->prevItem = freshNode;
    L->cursor->nextItem = freshNode;

    freshNode->prevItem = L->cursor;
    freshNode->nextItem = temp;
  }
}
// simply deletes the front element
void deleteFront(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  if (length(L) == 0) {
    printf("\nERROR: list length is zero\n");
    exit(1);
  }
  // if the cursor is on the head, set it to NULL
  if (L->cursor == L->head) {
    L->cursor = NULL;
  }
  // if length of list is 1, delete the entire list
  if (length(L) == 1) {
    clear(L);
  }
  // if not, just move head over to the right by one
  else {
    L->head->nextItem->prevItem = NULL;
    L->head = L->head->nextItem;
  }
}

// simply deletes the back element
void delteBack(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  if (length(L) == 0) {
    printf("\nERROR: list length is zero\n");
    exit(1);
  }
  // if the cursor is on the tail, set it to NULL
  if (L->cursor == L->tail) {
    L->cursor = NULL;
  }
  // if length of list is 1, delete the entire list
  if (length(L) == 1) {
    clear(L);
  }
  // if not, just move tail over to the left by one
  else {
    L->tail->prevItem->nextItem = NULL;
    L->tail = L->tail->prevItem;
  }
}

// deletes a node in the LIST
void delete (List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  if (length(L) <= 0 || L->cursor == NULL) {
    printf("\nERROR: Invalid List length or undefined cursor");
    exit(1);
  }
  // delete the links in the list
  Node point = L->cursor;
  // if cursor is on the head, follow the deleteFront procedure.
  if (L->cursor == L->head) {
    deleteFront(L);
  }
  // if cursor is on the tail, follow deleteBack procedure
  else if (L->cursor == L->tail) {
    deleteBack(L);
  }
  // otherwise, delete links normally
  else {
    L->cursor->prevItem->nextItem = point->nextItem;
    L->cursor->nextItem->prevItem = point->prevItem;
    L->cursor = NULL;
  }
}

// print the list
void printList(FILE *out, List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  Node temp = L->head;
  while (temp != NULL) {
    fprintf(out, "%d", temp->data);
    temp = temp->nextItem;
  }
}

// print the list but not in a file (just for tests)
void printList2(char *out, List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  Node temp = L->head;
  while (temp != NULL) {
    printf(out, "%d", temp->data);
    temp = temp->nextItem;
  }
}

// copies a list exactly except for the cursor
List copyList(List L) {
  if (L == NULL) {
    printf("\nERROR: Calling function on undefined List\n");
    exit(1);
  }
  List out = newList();
  Node copy = L->head;
  while (copy != NULL) {
    append(out, copy->data);
    copy = copy->nextItem;
  }
  return out;
}
